connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala

```diff
@@ -57,28 +69,105 @@ private[connect] object ErrorUtils extends Logging {
+
+      if (isTruncated) {
+        builder.setMessage(SparkConnectService.extractErrorMessage(error, MAX_MESSAGE_SIZE))
+      } else {
+        if (stackTraceInMessage) {
+          if (serverStackTraceEnabled || pySparkJVMStackTraceEnabled) {
+            builder.setMessage(
+              s"${error.getMessage}\n\n" +
+                s"JVM stacktrace:\n${ExceptionUtils.getStackTrace(error)}")
+          }
+        } else {
+          if (serverStackTraceEnabled) {
+            builder.addAllStackTrace(
+              error.getStackTrace
+                .map { stackTraceElement =>
+                  FetchErrorDetailsResponse.StackTraceElement
+                    .newBuilder()
+                    .setDeclaringClass(stackTraceElement.getClassName)
+                    .setMethodName(stackTraceElement.getMethodName)
+                    .setFileName(stackTraceElement.getFileName)
+                    .setLineNumber(stackTraceElement.getLineNumber)
+                    .build()
+                }
+                .toIterable
+                .asJava)
+          }
+        }
+      }
+
+      lastErrorOpt.foreach(builder.setCause(_))
+      lastErrorOpt = Some(builder.build())
+    }
+
+    val responseBuilder = FetchErrorDetailsResponse.newBuilder()
+
+    lastErrorOpt.foreach(responseBuilder.setExceptionInfo(_))
+
+    responseBuilder.build()
+  }
+
+  private def buildStatusFromThrowable(
+      st: Throwable,
+      enrichErrorEnabled: Boolean,
+      userId: String,
+      sessionId: String): RPCStatus = {
     val errorInfo = ErrorInfo
       .newBuilder()
       .setReason(st.getClass.getName)
       .setDomain("org.apache.spark")
-      .putMetadata(
-        "classes",
-        JsonMethods.compact(JsonMethods.render(allClasses(st.getClass).map(_.getName))))
-
-    lazy val stackTrace = Option(ExceptionUtils.getStackTrace(st))
-    val withStackTrace = if (stackTraceEnabled && stackTrace.nonEmpty) {
-      val maxSize = SparkEnv.get.conf.get(Connect.CONNECT_JVM_STACK_TRACE_MAX_SIZE)
-      errorInfo.putMetadata("stackTrace", StringUtils.abbreviate(stackTrace.get, maxSize))
-    } else {
-      errorInfo
+      .putMetadata("classes", serializeClasses(st))
+
+    if (enrichErrorEnabled) {
+      val errorId = UUID.randomUUID().toString
+
+      // The errorId of the exception.
+      errorInfo.putMetadata("errorId", errorId)
+
+      SparkConnectService
+        .getOrCreateIsolatedSession(userId, sessionId)
+        .errorIdToError
+        .put(errorId, st)
     }
 
     RPCStatus
       .newBuilder()
       .setCode(RPCCode.INTERNAL_VALUE)
-      .addDetails(ProtoAny.pack(withStackTrace.build()))
-      .setMessage(SparkConnectService.extractErrorMessage(st))
+      .addAllDetails(
+        Seq(errorInfo.build(), throwableToFetchErrorDetailsResponse(st))
+          .map(ProtoAny.pack(_))
+          .asJava)
+      .setMessage(SparkConnectService.extractErrorMessage(st, MAX_MESSAGE_SIZE))
```

Comment:
This is a change to the previous size of the message. We used to have 2048 bytes, now we have 512.

Reply:
I changed the message size back to 2048 in the splited pr.

Reply:
I undo the change